src/containers/projects/ProjectViewStateContainer.js

```diff
@@ -0,0 +1,51 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Obeo and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2
+ * which accompanies this distribution and is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *******************************************************************************/
+
+import React, { Component } from 'react';
+import { withRouter } from 'react-router-dom';
+
+import { ProjectView } from '../../components/projects/project/ProjectView';
+
+/**
+ * The ProjectViewStateContainerWithoutRouter is the stateful component used to
+ * manipulate the state of the ProjectView.
+ *
+ * It leverages the router API in order to retrieve the name of the project to
+ * display from the URL.
+ */
+class ProjectViewStateContainerWithoutRouter extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {};
+  }
+
+  async componentDidMount() {
+    try {
+      const { projectName } = this.props.match.params;
+      const jsonProjectResponse = await fetch(`/api/projects/${projectName}`);
+      const projectResponse = await jsonProjectResponse.json();
+
+      this.setState({
+        project: projectResponse
+      });
+    } catch (error) {
+      // To be handled later
+    }
+  }
+
+  render() {
+    const { project } = this.state;
+
+    if (project === undefined) {
+      return <p>Loading</p>;
+    }
+    return <ProjectView project={project} {...this.props} />;
+  }
+}
+export const ProjectViewStateContainer = withRouter(ProjectViewStateContainerWithoutRouter);
```

Comment:
That means ProjectViewStateContainerWithoutRouter will have access to the route ?

Reply:
ProjectViewStateContainerWithoutRouter is not exported, we only export the version decorated by "withRouter()". This usage of the high order component pattern (a pattern that I have presented during the training) will add several things in the props of the component.

The component will thus have access to some properties computed by the router like "match" (line 30) in order to retrieve the name of the project from the current URL. We can retrieve the value "projectName" from the match because we have specified this parameter in our URL in Main.js (you can also see it below).

```javascript
<main className={mainClassNames} {...props}>
  <Switch>
    <Route exact path="/" component={DashboardViewStateContainer} />
    <Route exact path="/projects" component={ListProjectsViewStateContainer} />
    <Route exact path="/projects/:projectName" component={ProjectViewStateContainer} />
  </Switch>
</main>
```